src/mlpack/core/tree/cellbound_impl.hpp

```diff
@@ -0,0 +1,994 @@
+ *
+ * @param centroid Vector which the centroid will be written to.
+ */
+template<typename MetricType, typename ElemType>
+inline void CellBound<MetricType, ElemType>::Center(
+    arma::Col<ElemType>& center) const
+{
+  // Set size correctly if necessary.
+  if (!(center.n_elem == dim))
+    center.set_size(dim);
+
+  for (size_t i = 0; i < dim; i++)
+    center(i) = bounds[i].Mid();
+}
+
+template<typename MetricType, typename ElemType>
+template<typename MatType>
+void CellBound<MetricType, ElemType>::AddBound(
+    const arma::Col<ElemType>& loCorner,
+    const arma::Col<ElemType>& hiCorner,
+    const MatType& data)
+{
+  assert(numBounds < loBound.n_cols);
+  assert(loBound.n_rows == dim);
+  assert(loCorner.n_elem == dim);
+  assert(hiCorner.n_elem == dim);
+
+  for (size_t k = 0; k < dim; k++)
+  {
+    loBound(k, numBounds) = std::numeric_limits<ElemType>::max();
+    hiBound(k, numBounds) = std::numeric_limits<ElemType>::lowest();
+  }
+
+  for (size_t i = 0; i < data.n_cols; i++)
+  {
+    size_t k = 0;
+    // Check if the point is contained in the hyperrectangle.
+    for (k = 0; k < dim; k++)
+      if (data(k, i) < loCorner[k] || data(k, i) > hiCorner[k])
+        break;
+
+    if (k < dim)
+      continue; // The point is not contained in the hyperrectangle.
+
+    // Srink the bound.
+    for (k = 0; k < dim; k++)
+    {
+      loBound(k, numBounds) = std::min(loBound(k, numBounds), data(k, i));
+
+      hiBound(k, numBounds) = std::max(hiBound(k, numBounds), data(k, i));
+    }
+  }
+
+  for (size_t k = 0; k < dim; k++)
+    if (loBound(k, numBounds) > hiBound(k, numBounds))
+      return; // The hyperrectangle does not contain points.
+
+  numBounds++;
+}
+
+
+template<typename MetricType, typename ElemType>
+template<typename MatType>
+void CellBound<MetricType, ElemType>::InitHighBound(size_t numEqualBits,
+                                                    const MatType& data)
+{
+  arma::Col<AddressElemType> tmpHiAddress(hiAddress);
+  arma::Col<AddressElemType> tmpLoAddress(hiAddress);
+  arma::Col<ElemType> loCorner(tmpHiAddress.n_elem);
+  arma::Col<ElemType> hiCorner(tmpHiAddress.n_elem);
+
+  assert(tmpHiAddress.n_elem > 0);
+
+  // We have to calculate the number of subrectangles since the maximum number
+  // of hyperrectangles is restricted.
+  size_t numCorners = 0;
+  for (size_t pos = numEqualBits + 1; pos < order * tmpHiAddress.n_elem; pos++)
+  {
+    size_t row = pos / order;
+    size_t bit = order - 1 - pos % order;
+
+    // This hyperrectangle is not contained entirely in the bound.
+    // So, the number of hyperrectangles should be increased.
+    if (tmpHiAddress[row] & ((AddressElemType) 1 << bit))
+      numCorners++;
+
+    // We ran out of the limit of hyperrectangles. In that case we enlare
+    // the last hyperrectangle.
+    if (numCorners >= maxNumBounds / 2)
+      tmpHiAddress[row] |= ((AddressElemType) 1 << bit);
```

Comment:
I think the first strategy is best here.  I don't think it's a problem if the functions are not symmetric, and I think the performance impact will be negligible.  Definitely the first strategy is best, but if you want to make an extra pass that's fine too.  I just want to avoid a situation where a user says "but if I set maxNumBounds to 8 and 9 the performance is the same, and that doesn't make sense!"